### ISSUE TITLE
[Bug]: transform keyword in ax.plot(...)

### ISSUE THREAD

**Issue by kovalp** [matplotlib/matplotlib#21008]
### Bug summary

The keyword argument `transform` of the method ax.plot(...) malfunctions with size of the samples greater than 1000.

For example, the following script

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.transforms as transforms


ax = plt.gca()

xx = np.linspace(0.0, 4 * np.pi, num=1001)
yy = np.cos(xx)

t = transforms.Affine2D().translate(10.0, 0.0).scale(2.0, 0.5)

lines = ax.plot(xx, yy, transform=t + ax.transData)

plt.show()
```

Produces an expected figure

![Figure_1000](https://user-images.githubusercontent.com/8431900/132308652-075bd5b6-6d14-4eb9-8615-aeef15b1a737.png)

with `num=1000`

and an empty figure

![Figure_1001](https://user-images.githubusercontent.com/8431900/132316353-90db4ede-cf60-4a5e-a850-d53f8e16047a.png)

with `num=1001`.

### Code for reproduction

```python
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.transforms as transforms


ax = plt.gca()

xx = np.linspace(0.0, 4 * np.pi, num=1001)
yy = np.cos(xx)

t = transforms.Affine2D().translate(10.0, 0.0).scale(2.0, 0.5)

lines = ax.plot(xx, yy, transform=t + ax.transData)

plt.show()
```

### Actual outcome

![Figure_1001](https://user-images.githubusercontent.com/8431900/132316487-226e2f0b-3c6a-4409-a70b-11eb256b6b26.png)

### Expected outcome

![Figure_1000](https://user-images.githubusercontent.com/8431900/132316499-b5fbb3f8-d5ce-4200-ab5e-d84a97183c87.png)

### Operating system

Linux

### Matplotlib Version

3.4.3

### Matplotlib Backend

Qt5Agg

### Python version

Python 3.8.7

### Installation

pip

**Comment by jklymak** [comment]
I agree this seems like a bug.  It is very strange in that `scatter` works fine, so it is something in how `plot` is applying the transform

**Comment by jklymak** [comment]
This is due to https://github.com/matplotlib/matplotlib/blob/8ce14fb60d9441ce3fb92e6e59b4aa3146aca6c1/lib/matplotlib/lines.py#L663
causing a slicing of the data if the size is greater than 1000 (hard coded!) and then the slicing only happening in the view window, which has been transformed.  i.e. your _data_ goes from 0 to 4*pi, but the window goes from 20 to 20 + 4*pi.  So all the data is sliced away, and nothing is plotted.  

I'm not sure what the todo is here.  My tendency would be to just remove the slicing under the hope that modern machines don't need this optimization.  As you can see it is only applied in a very limited set of circumstances.  Conversely, one could (should) apply the transform to the data first, but that is a bit ugly.  

This optimization was last touched 12 years ago.

**Comment by kovalp** [comment]
I am transforming the data manually now. However, my idea when using
transforms would be in not duplicating the data, of course.

**Comment by anntzer** [comment]
I would just add a check to restrict the optimization to the case where `self.get_transform() == self.axes.transData`, I guess.

**Comment by jklymak** [comment]
Yeah that's reasonable as well. OTOH it's a pretty specialized optimization.

**Comment by kovalp** [comment]
> This is due to
> 
> https://github.com/matplotlib/matplotlib/blob/8ce14fb60d9441ce3fb92e6e59b4aa3146aca6c1/lib/matplotlib/lines.py#L663

Gosh, this is deep. The condition to allow for the optimization turned out not sufficiently excluding after all.

When I found this bug, the outcome of the plotting was not obviously wrong. Namely, I used the transform to account for the units. When up-scaling (smaller units), the plot appeared as expected, but when down-scaling, the plot was visible, but smaller that the original and cut.

This making me think that I am not getting some matplotlib argument right. However, when the 1000 threshold appeared, things become more certain.
